### PR TITLE
Implement HashBagWithHashingStrategy. Closes #23.

### LINKED ISSUE #23: Implement Bags that use hashing strategies.
`UnifiedSet` has a variant called `UnifiedSetWithHashingStrategy` that uses a `HashingStrategy` instead of using `equals()` and `hashCode()`.

Similarly, `HashBag` should have a a variant that uses a `HashingStrategy`.

### PR DESCRIPTION
**Pull request by bhav0904**
Signed-off-by: Bhavana Hindupur <bhavana.hindupur@gmail.com>

### CODE REVIEW

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/bag/strategy/mutable/HashBagWithHashingStrategy.java [hunk 1]
```diff
@@ -0,0 +1,531 @@
...
+import org.eclipse.collections.impl.bag.mutable.AbstractMutableBag;
+import org.eclipse.collections.impl.block.factory.primitive.IntToIntFunctions;
+import org.eclipse.collections.impl.map.mutable.UnifiedMap;
+import org.eclipse.collections.impl.map.mutable.primitive.ObjectIntHashMapWithHashingStrategy;
+import org.eclipse.collections.impl.multimap.bag.HashBagMultimap;
+import org.eclipse.collections.impl.utility.ArrayIterate;
+import org.eclipse.collections.impl.utility.Iterate;
+
+public class HashBagWithHashingStrategy<T>
+        extends AbstractMutableBag<T>
```
Comment (motlin) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/bag/strategy/mutable/HashBagWithHashingStrategy.java]: Several methods are identical between `HashBag` and `HashBagWithHashingStrategy`. What do you think about extracting a common superclass above both classes?

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/bag/strategy/mutable/HashBagWithHashingStrategy.java [hunk 2]
```diff
@@ -0,0 +1,531 @@
...
+        implements Serializable
+{
+    private static final long serialVersionUID = 1L;
+
+    private MutableObjectIntMap<T> items;
+    private int size;
+
+    private HashingStrategy<? super T> hashingStrategy;
```
Comment (motlin) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/bag/strategy/mutable/HashBagWithHashingStrategy.java]: The field `hashingStrategy` can be final, probably `items` too.

File: unit-tests/src/test/java/org/eclipse/collections/impl/bag/strategy/mutable/HashBagWithHashingStrategyTest.java [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+import org.eclipse.collections.impl.test.domain.Person;
+import org.junit.Assert;
+import org.junit.Test;
+
+/**
+ * JUnit test suite for {@link HashBagWithHashingStrategy}.
+ */
+public class HashBagWithHashingStrategyTest extends MutableBagTestCase
+{
+    private static final HashingStrategy<String> STRING_HASHING_STRATEGY = HashingStrategies.nullSafeHashingStrategy(new HashingStrategy<String>()
```
Comment (motlin) [inline unit-tests/src/test/java/org/eclipse/collections/impl/bag/strategy/mutable/HashBagWithHashingStrategyTest.java]: Both constant hashing strategies are the same as each other, and the same as `HashingStrategies.defaultStrategy()`. The other strategies for first name and last name allow for easy collisions, which is good for testing.

**Comment by motlin** [comment]
Looks really good overall, thanks!

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/bag/strategy/mutable/HashBagWithHashingStrategy.java (hunk 1 above)
Comment (nikhilnanivadekar) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/bag/strategy/mutable/HashBagWithHashingStrategy.java]: There is something similar for UnifiedSet and UnifiedSetWithHashingStrategy: [AbstractUnifiedSet](https://github.com/eclipse/eclipse-collections/blob/8d45bc751e796c1e0f8cc21a26d4f96b060392c2/eclipse-collections/src/main/java/org/eclipse/collections/impl/set/AbstractUnifiedSet.java)

**Comment by bhav0904** [comment]
Worked on all the review comments.

**Comment by itohro** [comment]
The code changes look good to me.

@bhav0904, could you please address 2 things before we merge? 

- Squash 2 commits (1055855 and 0a28d42) into 1 commit.
- Rebase on top of master.

**Pull request closed without merging**